### PR TITLE
[WIP] GradTTS Implementation

### PR DESCRIPTION
**Pull request by Arya-Li**
## What?
Add GradTTS implementation

## Why?
to add gradient-based diffusion modeling to the TTS task

## See also
https://github.com/espnet/espnet/discussions/5527
https://github.com/huawei-noah/Speech-Backbones/tree/main/Grad-TTS

### CODE REVIEW

File: espnet2/tts/gradtts/denoiser.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
+import math
```
Comment (ftshijt) [inline espnet2/tts/gradtts/denoiser.py:1]: Please include the source reference here to credit the original implementation.

File: espnet2/tts/gradtts/denoiser.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+import torch
+from einops import rearrange
+
+from espnet2.tts.abs_tts import AbsTTS
```
Comment (ftshijt) [inline espnet2/tts/gradtts/denoiser.py:6]: no need to use abstts for every module, torch.nn.Module would be enough

File: espnet2/tts/gradtts/denoiser.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+
+class Mish(AbsTTS):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/denoiser.py:9]:
```suggestion
class Mish(torch.nn.Module):
```
Please change accordingly.

File: espnet2/tts/gradtts/gradtts.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
+from espnet2.tts.abs_tts import AbsTTS
+from espnet2.tts.gradtts.utils import (
+    duration_loss,
+    fix_len_compatibility,
+    generate_path,
+    sequence_mask,
+)
+
+
+class LayerNorm(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/gradtts.py:10]: Is there any difference between this layernorm and the torch's implementation? It seems similar to me

File: espnet2/tts/gradtts/gradtts.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+        variance = torch.mean((x - mean) ** 2, 1, keepdim=True)
+
+        x = (x - mean) * torch.rsqrt(variance + self.eps)
+
+        shape = [1, -1] + [1] * (n_dims - 2)
+        x = x * self.gamma.view(*shape) + self.beta.view(*shape)
+        return x
+
+
+class ConvReluNorm(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/gradtts.py:31]: what is the basemodule here (where is it defined?)

File: espnet2/tts/gradtts/gradtts.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        x_org = x
+        for i in range(self.n_layers):
+            x = self.conv_layers[i](x * x_mask)
+            x = self.norm_layers[i](x)
+            x = self.relu_drop(x)
+        x = x_org + self.proj(x)
+        return x * x_mask
+
+
+class DurationPredictor(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/gradtts.py:66]: This part is similar to fastspeech's. Any differences? If not big different, would recommend use that one instead.

File: espnet2/tts/gradtts/gradtts.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        x = self.drop(x)
+        x = self.conv_2(x * x_mask)
+        x = torch.relu(x)
+        x = self.norm_2(x)
+        x = self.drop(x)
+        x = self.proj(x * x_mask)
+        return x * x_mask
+
+
+class MultiHeadAttention(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/gradtts.py:95]: It would be better to use ESPnet's multihead attention implementation for consistency.

File: espnet2/tts/gradtts/gradtts.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        x_final = x_flat.view([batch, heads, length, 2 * length])[:, :, :, 1:]
+        return x_final
+
+    def _attention_bias_proximal(self, length):
+        r = torch.arange(length, dtype=torch.float32)
+        diff = torch.unsqueeze(r, 0) - torch.unsqueeze(r, 1)
+        return torch.unsqueeze(torch.unsqueeze(-torch.log1p(torch.abs(diff)), 0), 0)
+
+
+class FFN(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/gradtts.py:217]: Ditto, we have similar definition in ESPnet

File: espnet2/tts/gradtts/gradtts.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+            y = self.drop(y)
+            x = self.norm_layers_1[i](x + y)
+            y = self.ffn_layers[i](x, x_mask)
+            y = self.drop(y)
+            x = self.norm_layers_2[i](x + y)
+        x = x * x_mask
+        return x
+
+
+class TextEncoder(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/gradtts.py:280]: Did not see significant change, might easier to directly use Fastspeech's text encoder?

File: espnet2/tts/gradtts/utils.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import torch
+
+
+def sequence_mask(length, max_length=None):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/utils.py:4]: might be better to use espnet torch_utils functions for the masking to keep the consistency

**Review by ftshijt — COMMENTED** [review]
Many thanks for your initial commit. I raise up a few notes or comments related to the design-level suggestions. Mostly for consistency issues, probably better to follow them after you make sure that the model can be working in the espnet trainer, which makes it easier to locate the early stage bugs.

File: espnet2/tts/gradtts/gradtts.py (hunk 5 above)
Comment (Arya-Li) [reply espnet2/tts/gradtts/gradtts.py]: This is defined in the gradtts source code, and I just added it. I want to try migrating it as a whole and try to get through the code first.

File: espnet2/tts/gradtts/gradtts.py (hunk 9 above)
Comment (Arya-Li) [reply espnet2/tts/gradtts/gradtts.py]: Okay, I'll try changing it.

File: espnet2/tts/gradtts/base.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# MIT License for more details.
+
+import numpy as np
+import torch
+
+
+class BaseModule(torch.nn.Module):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/base.py:13]: We do not necessarily need this (as all the functions will be taken care during our espnet trainer). My suggestion would to simple use torch.nn.Module as base class.

File: espnet2/tts/gradtts/denoiser.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
+import math
+
+import torch
+from einops import rearrange
+
+from espnet2.tts.gradtts.base import BaseModule
```
Comment (ftshijt) [inline espnet2/tts/gradtts/denoiser.py:6]:
```suggestion
```

File: espnet2/tts/gradtts/denoiser.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+
+
+class Mish(BaseModule):
```
Comment (ftshijt) [inline espnet2/tts/gradtts/denoiser.py:9]:
```suggestion
class Mish(torch.nn.Module):
```
As suggested, use torch.nn.Module
